File: internal/model/subscription.go
package model

import "time"

type BillingSource string

const (
	BillingSourceSubscription BillingSource = "subscription"
	BillingSourceBalance      BillingSource = "balance"
)

type WindowMode string

const (
	WindowModeFixed   WindowMode = "fixed"
	WindowModeSliding WindowMode = "sliding"
)

type LimitType string

const (
	LimitTypeDaily     LimitType = "daily"
	LimitTypeWeekly    LimitType = "weekly"
	LimitTypeMonthly   LimitType = "monthly"
	LimitTypeRolling5h LimitType = "rolling_5h"
	LimitTypeTotal     LimitType = "total"
)

type SubscriptionStatus string

const (
	SubscriptionStatusActive    SubscriptionStatus = "active"
	SubscriptionStatusPaused    SubscriptionStatus = "paused"
	SubscriptionStatusExpired   SubscriptionStatus = "expired"
	SubscriptionStatusCancelled SubscriptionStatus = "cancelled"
)

type SubscriptionPlan struct {
	ID          string    `json:"id"`
	Name        string    `json:"name"`
	Description string    `json:"description"`
	Enabled     bool      `json:"enabled"`
	CreatedAt   time.Time `json:"createdAt"`
	UpdatedAt   time.Time `json:"updatedAt"`
}

type SubscriptionPlanLimit struct {
	ID          string     `json:"id"`
	PlanID      string     `json:"planId"`
	LimitType   LimitType  `json:"limitType"`
	WindowMode  WindowMode `json:"windowMode"`
	LimitMicros int64      `json:"limitMicros"`
	CreatedAt   time.Time  `json:"createdAt"`
	UpdatedAt   time.Time  `json:"updatedAt"`
}

type UserSubscription struct {
	ID        string             `json:"id"`
	UserID    string             `json:"userId"`
	PlanID    string             `json:"planId"`
	StartsAt  time.Time          `json:"startsAt"`
	ExpiresAt *time.Time         `json:"expiresAt"`
	Status    SubscriptionStatus `json:"status"`
	CreatedAt time.Time          `json:"createdAt"`
	UpdatedAt time.Time          `json:"updatedAt"`
}

type UserBillingSetting struct {
	UserID          string        `json:"userId"`
	PrimarySource   BillingSource `json:"primarySource"`
	SecondarySource BillingSource `json:"secondarySource"`
	CreatedAt       time.Time     `json:"createdAt"`
	UpdatedAt       time.Time     `json:"updatedAt"`
}

type BillingEvent struct {
	ID                 string        `json:"id"`
	RequestLogID       *string       `json:"requestLogId"`
	UserID             string        `json:"userId"`
	UserSubscriptionID *string       `json:"userSubscriptionId"`
	Source             BillingSource `json:"source"`
	EventType          string        `json:"eventType"`
	AmountMicros       int64         `json:"amountMicros"`
	CreatedAt          time.Time     `json:"createdAt"`
}

// --- Request / Response DTOs ---

type SubscriptionPlanRequest struct {
	Name        string            `json:"name" binding:"required,min=1,max=64"`
	Description string            `json:"description" binding:"max=256"`
	Enabled     bool              `json:"enabled"`
	Limits      []PlanLimitRequest `json:"limits"`
}

type PlanLimitRequest struct {
	LimitType   LimitType  `json:"limitType" binding:"required"`
	WindowMode  WindowMode `json:"windowMode" binding:"required"`
	LimitMicros int64      `json:"limitMicros" binding:"required,min=0"`
}

type SubscriptionPlanResponse struct {
	ID          string                  `json:"id"`
	Name        string                  `json:"name"`
	Description string                  `json:"description"`
	Enabled     bool                    `json:"enabled"`
	Limits      []SubscriptionPlanLimit `json:"limits"`
	CreatedAt   time.Time               `json:"createdAt"`
	UpdatedAt   time.Time               `json:"updatedAt"`
}

type AssignSubscriptionRequest struct {
	PlanID    string     `json:"planId" binding:"required"`
	ExpiresAt *time.Time `json:"expiresAt"`
}

type UserSubscriptionResponse struct {
	ID        string                  `json:"id"`
	UserID    string                  `json:"userId"`
	PlanID    string                  `json:"planId"`
	PlanName  string                  `json:"planName"`
	StartsAt  time.Time               `json:"startsAt"`
	ExpiresAt *time.Time              `json:"expiresAt"`
	Status    SubscriptionStatus      `json:"status"`
	Limits    []SubscriptionPlanLimit `json:"limits"`
	CreatedAt time.Time               `json:"createdAt"`
	UpdatedAt time.Time               `json:"updatedAt"`
}

type WindowRemaining struct {
	LimitType   LimitType `json:"limitType"`
	WindowMode  WindowMode `json:"windowMode"`
	LimitMicros int64     `json:"limitMicros"`
	UsedMicros  int64     `json:"usedMicros"`
	LeftMicros  int64     `json:"leftMicros"`
}

type BillingStateResponse struct {
	BalanceMicros int64                    `json:"balanceMicros"`
	BalanceUsd    string                   `json:"balanceUsd"`
	Subscription  *UserSubscriptionResponse `json:"subscription"`
	Windows       []WindowRemaining        `json:"windows"`
	PrimarySource BillingSource            `json:"primarySource"`
	SecondarySource BillingSource          `json:"secondarySource"`
}

type UpdateBillingPriorityRequest struct {
	PrimarySource BillingSource `json:"primarySource" binding:"required"`
}

File: web/src/api/subscription.ts
const API_BASE = '/api'

function getAuthHeaders() {
  const token = localStorage.getItem('token')
  return {
    'Content-Type': 'application/json',
    Authorization: `Bearer ${token}`,
  }
}

export type LimitType = 'daily' | 'weekly' | 'monthly' | 'rolling_5h' | 'total'
export type WindowMode = 'fixed' | 'sliding'
export type SubscriptionStatus = 'active' | 'paused' | 'expired' | 'cancelled'

export interface SubscriptionPlanLimit {
  id: string
  planId: string
  limitType: LimitType
  windowMode: WindowMode
  limitMicros: number
  createdAt: string
  updatedAt: string
}

export interface SubscriptionPlanResponse {
  id: string
  name: string
  description: string
  enabled: boolean
  limits: SubscriptionPlanLimit[]
  createdAt: string
  updatedAt: string
}

export interface PlanLimitRequest {
  limitType: LimitType
  windowMode: WindowMode
  limitMicros: number
}

export interface SubscriptionPlanRequest {
  name: string
  description: string
  enabled: boolean
  limits: PlanLimitRequest[]
}

export interface UserSubscriptionResponse {
  id: string
  userId: string
  planId: string
  planName: string
  startsAt: string
  expiresAt: string | null
  status: SubscriptionStatus
  limits: SubscriptionPlanLimit[]
  createdAt: string
  updatedAt: string
}

// Plan CRUD
export async function getPlans(): Promise<SubscriptionPlanResponse[]> {
  const res = await fetch(`${API_BASE}/admin/subscriptions/plans`, {
    headers: getAuthHeaders(),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '获取套餐列表失败')
  }
  const data = await res.json()
  return data.plans || []
}

export async function createPlan(req: SubscriptionPlanRequest): Promise<SubscriptionPlanResponse> {
  const res = await fetch(`${API_BASE}/admin/subscriptions/plans`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify(req),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '创建套餐失败')
  }
  return res.json()
}

export async function getPlan(id: string): Promise<SubscriptionPlanResponse> {
  const res = await fetch(`${API_BASE}/admin/subscriptions/plans/${id}`, {
    headers: getAuthHeaders(),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '获取套餐详情失败')
  }
  return res.json()
}

export async function updatePlan(id: string, req: SubscriptionPlanRequest): Promise<SubscriptionPlanResponse> {
  const res = await fetch(`${API_BASE}/admin/subscriptions/plans/${id}`, {
    method: 'PUT',
    headers: getAuthHeaders(),
    body: JSON.stringify(req),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '更新套餐失败')
  }
  return res.json()
}

export async function deletePlan(id: string): Promise<void> {
  const res = await fetch(`${API_BASE}/admin/subscriptions/plans/${id}`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '删除套餐失败')
  }
}

export async function setPlanEnabled(id: string, enabled: boolean): Promise<void> {
  const res = await fetch(`${API_BASE}/admin/subscriptions/plans/${id}/enabled`, {
    method: 'PATCH',
    headers: getAuthHeaders(),
    body: JSON.stringify({ enabled }),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '更新套餐状态失败')
  }
}

// User subscription management
export async function getUserSubscription(userId: string): Promise<UserSubscriptionResponse | null> {
  const res = await fetch(`${API_BASE}/admin/users/${userId}/subscription`, {
    headers: getAuthHeaders(),
  })
  if (!res.ok) {
    if (res.status === 404) return null
    const data = await res.json()
    throw new Error(data.error || '获取用户订阅失败')
  }
  const data = await res.json()
  return data.subscription || null
}

export async function assignSubscription(userId: string, req: { planId: string; expiresAt?: string }): Promise<UserSubscriptionResponse> {
  const res = await fetch(`${API_BASE}/admin/users/${userId}/subscription`, {
    method: 'POST',
    headers: getAuthHeaders(),
    body: JSON.stringify(req),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '分配订阅失败')
  }
  return res.json()
}

export async function cancelSubscription(userId: string): Promise<void> {
  const res = await fetch(`${API_BASE}/admin/users/${userId}/subscription`, {
    method: 'DELETE',
    headers: getAuthHeaders(),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '取消订阅失败')
  }
}

export async function updateSubscriptionExpiry(userId: string, expiresAt: string): Promise<void> {
  const res = await fetch(`${API_BASE}/admin/users/${userId}/subscription`, {
    method: 'PATCH',
    headers: getAuthHeaders(),
    body: JSON.stringify({ expiresAt }),
  })
  if (!res.ok) {
    const data = await res.json()
    throw new Error(data.error || '更新到期时间失败')
  }
}
